### PR TITLE
Fixes DF-1601 - fix incomplete filtering

### LINKED ISSUE #261: Odd filter behavior when only one date is chosen
When using the filter on pages such as /blog/ and /newsroom/, if only a single date is filled out, after submission that entered date will switch to the other date field (To -> From or visa versa).

To reproduce: on /newsroom/, enter a To: month and year, but enter nothing for From.  Hit submit.  Now look at the filter.   The month and year you selected will be in the From section.

I believe this has to do with some clever behavior that swaps the From and To dates if the To date is earlier than the From date.

### PR DESCRIPTION
**Pull request by anselmbradford**
Fixes behavior of filters when partial date ranges are entered.
Fixes #261 and #365.

## Additions

- Adds `formatDateString` function for formatting the values returned from the date inputs.
- Adds jsdoc comment to `isValidDateRange`.

## Changes

- Makes the date range displayed for results only display as a single month + year when both start and end date range are the same.
- Changes `validDateRange` function name to `isValidDateRange` to better convey it returns a boolean.

## Testing

- Visit /blog/ or /newsroom/ and play around with the dates in the filtering.

## Review

- @jimmynotjim 
- @sebworks 
- @Scotchester 

## Preview

filters

## Notes

- The swapping behavior seems kind of off to me. I'm wondering if perhaps more ideally the user would be prompted that there's an error in how they entered the range and whether they'd like to swap the values. That's out of scope for this PR however.

### CODE REVIEW

File: static/js/main.js [hunk 1]
```diff
@@ -13245,21 +13245,58 @@ $('.js-validate-filters').each(function() {
...
-        var validDate = validDateRange(
-            new Date( Date.parse($gte.val()) ),
-            new Date( Date.parse($lte.val()) )
+        var gteVal = $gte.val();
+        var lteVal = $lte.val();
+
+        var gteParsedVal = formatDateString(gteVal, lteVal);
+        var lteParsedVal = formatDateString(lteVal, gteVal);
+
+        var validDate = isValidDateRange(
```
Comment (jimmynotjim) [inline static/js/main.js]: Might rename this `var validRange` for clarity on what you're testing against on [L:13293](https://github.com/cfpb/cfgov-refresh/pull/366/files#diff-93471d4bf84d0569fadcebe5fba72773R13293)

File: static/js/app.js [hunk 2]
```diff
@@ -430,21 +430,58 @@ $('.js-validate-filters').each(function() {
...
-        var validDate = validDateRange(
-            new Date( Date.parse($gte.val()) ),
-            new Date( Date.parse($lte.val()) )
+        var gteVal = $gte.val();
+        var lteVal = $lte.val();
+
+        var gteParsedVal = formatDateString(gteVal, lteVal);
+        var lteParsedVal = formatDateString(lteVal, gteVal);
+
+        var validDate = isValidDateRange(
```
Comment (jimmynotjim) [inline static/js/app.js]: Might rename this `var validRange` for clarity on what you're testing against on [L:478](https://github.com/cfpb/cfgov-refresh/pull/366/files#diff-6d7870f99d3f759dc8b9f7a6969f1129R478)

*Sorry for the double comment, realized I was commenting on the compiled file

Comment (anselmbradford) [inline static/js/app.js]: Would `var validDateRange` do? Went with that for consistency of the name with the `isValidDateRange` function. Fixed in https://github.com/cfpb/cfgov-refresh/commit/38bfb333d8cac48ed4b8dad95fba3d4524ea81e6

**Comment by jimmynotjim** [comment]
I think auto swapping the dates if the user inverted them is helpful UX. Why make them have to stop and think about what they meant to do instead of just fixing it for them? It's one of those things no one will really notice but appreciate if they knew it was handled for them.

**Comment by jimmynotjim** [comment]
LTGM

**Comment by anselmbradford** [comment]
> I think auto swapping the dates if the user inverted them is helpful UX. Why make them have to stop and think about what they meant to do instead of just fixing it for them? It's one of those things no one will really notice but appreciate if they knew it was handled for them.

Yeah, no doubt this was done for that reason. Thinking about it, here's what bothers me about it: There's no reason the date range couldn't mean "from a more present date to one in the past". Take this range:

screen shot 2015-03-16 at 10 43 50 am

I'd kind of expect that the posts would be returned in reverse chronological order from Feb 2011 to Jan 2011, which is what happens. _However_, the swapped values make it look like it should be displaying posts in chronological order, starting from January 2011:

screen shot 2015-03-16 at 10 52 53 am

So that's essentially says "display posts from January 2011 to February 2011" ... but what's actually displayed in posts from February 2011 in reverse chronological order back to January 2011.

File: static/js/app.js (hunk 2 above)
Comment (jimmynotjim) [inline static/js/app.js]: Yep, just as long as it's not the date itself you're validating (that's what I thought it was while quickly skimming until I went back up to see where it was set)

**Comment by jimmynotjim** [comment]
Ah, I didn't notice the label was reversed. I definitely agree that part should be fixed.

**Comment by Scotchester** [comment]
That is, in fact, exactly why we did the swapping behavior. To play a little devil's advocate:

1. Does anyone ever expect blog/news posts to be returned in chronological order?
2. This is a filtering function, not a sorting function. I was surprised to hear your interpretation of "from/to" as "top-to-bottom of the results". I would expect most users to only take it to mean "I want results form this month to that month". But you know what they say about assumptions. Probably something worth doing some usability testing on.

**Comment by jimmynotjim** [comment]
1. It depends, but if I'm filtering by date without a keyword to match, I kind of expect them to be sorted by date as well.
2. Since the API does return them chronologically, we may want to include a label so it's obvious or add a sorting option (even if it's just simply "sort by oldest/newest").

**Comment by anselmbradford** [comment]
Thanks for the input @Scotchester, hope it doesn't feel like I'm trashing the logic you guys put into the project! :)
Fundamentally, I'd kind of expect the interface to do what I told it to do. If I can put in a date that was not allowed, I'd kind of prefer the interface didn't allow me to enter erroneous info in the first place.

**Comment by Scotchester** [comment]
Doesn't feel like that at all :) Our decisions are always up for constructive criticism.

One thing I meant to say in my previous comment that didn't: I agree that it doesn't need to be swapped at all. The result set is the same (ordering aside) if the from and to dates are swapped, so let's not confuse the user by swapping them.

On your second paragraph, Ans: agreed. I think we always intended to have better client-side validation "when we get around to it". We should prevent users from entering bad/impossible selections, when possible.

**Comment by anselmbradford** [comment]
Moved to https://github.com/cfpb/cfgov-refresh/issues/416

**Pull request merged**